### PR TITLE
Allow white space in FullNameSeparator fields

### LINKED ISSUE #5074: Whitespace FullNameSeparator for TreeDefItems set to NULL when TreeDefItem is updated
**Describe the bug**
The Specify 7 frontend trims whitespace before sending the values to the backend. 
This is a big problem for TreeDefItem records, where a tab, space, or other whitespace character should be allowed. 
Furthermore, Specify trims any existing whitespace value when the record is saved. In this case, setting the fullNameSeparator of the TaxonTreeDefItem to NULL. 

**To Reproduce**

1.  Identify any TreeDefItem record which has a non-NULL whitespace (such as a space) as a fullNameSeparator
2. Make any edit to the TreeDefItem's TreeDef or to the TreeDefItem (does not have to be to the fullNameSeparator field)
3. Make any edit to a Tree node at the rank to regenerate the node's fullName
4. See that the names of the Tree nodes for each rank included in the fullname are concatenated and  the fullName is constructed without any whitespace 

https://github.com/specify/specify7/assets/64045831/4ff5a722-764d-4b7f-b2a8-fa8998884cc1

Please, also fill out the following information manually:

- OS: macOS
- Browser: Chrome
- Specify 7 Version: 7.9.7 development (edge - https://github.com/specify/specify7/commit/080b32ceaec14d747291bbb020d6329f48dcedc0)

### PR DESCRIPTION
**Pull request by melton-jason**
Fixes #5074 

While this covers the case of fullNameSeparator fields for TreeDefItems, there is always the possibility that there may be other fields which would need to covered as well. A general solution was not included in this PR for the sake of complexity, but one may be desired.  

Always ensures that the following fields are not trimmed:
- TaxonTreeDefItem -> fullNameSeparator
- GeographyTreeDefItem -> fullNameSeparator
- StorageTreeDefItem -> fullNameSeparator
- GeologicTimePeriodTreeDefItem -> fullNameSeparator
- LithoStratTreeDefItem -> fullNameSeparator

Also adds support for a `whiteSpaceSensitive` initialize attribute for `uitype=text` fields on form definitions, which tells the frontend that any value in that field should not be trimmed (no leading and trailing whitespace should be removed). 

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [x] Add automated tests
- [x] Add relevant issue to release milestone

### Testing instructions

**fullNameSeparator**
- Make any edit to a TreeDefItem which is included in the fullName and save
  - For example, if Genus is edited then prior to this PR any `Genus Species` Taxon records would become `GenusSpecies` 
- In the TreeViewer or QueryBuilder, identify a node at a low level (Species, SubSpecies, etc.)
- [ ] Ensure the Full Name of the tree node was constructed as expected

(For a video demonstrating the Issue see https://github.com/specify/specify7/issues/5074#issue-2393357073)

**whiteSpaceSensitive form attribute**
- Identify any non-relationship field on any table which is of type `text` or `String`
- On the Form Definition for the table, add the field to the definition (if needed) and add `whiteSpaceSensitive=true` to the initialize attributes of the field (example label and cell provided below)
- [ ] Navigate to a DataEntry form of the table and enter a value which would normally be trimmed by Specify (such as `    test   `) and ensure the value remains as-is after the form is saved

### CODE REVIEW

**Review by emenslin — CHANGES_REQUESTED** [review]
### Testing instructions

**fullNameSeparator**
- Make any edit to a TreeDefItem which is included in the fullName and save
  - For example, if Genus is edited then prior to this PR any `Genus Species` Taxon records would become `GenusSpecies` 
- In the TreeViewer or QueryBuilder, identify a node at a low level (Species, SubSpecies, etc.)
- [x] Ensure the Full Name of the tree node was constructed as expected


Looks good, was able to recreate the original issue and fix it in this pr 


**whiteSpaceSensitive form attribute**
- Identify any non-relationship field on any table which is of type `text` or `String`
- On the Form Definition for the table, add the field to the definition (if needed) and add `whiteSpaceSensitive=true` to the initialize attributes of the field (example label and cell provided below)
- [ ] Navigate to a DataEntry form of the table and enter a value which would normally be trimmed by Specify (such as `    test   `) and ensure the value remains as-is after the form is saved

This does not work, the value still gets trimmed 

https://github.com/specify/specify7/assets/150940840/642af47a-9c7b-4a52-b2c2-e84334a5db36

**Review by emenslin — APPROVED** [review]
### Testing instructions

**fullNameSeparator**
- Make any edit to a TreeDefItem which is included in the fullName and save
  - For example, if Genus is edited then prior to this PR any `Genus Species` Taxon records would become `GenusSpecies` 
- In the TreeViewer or QueryBuilder, identify a node at a low level (Species, SubSpecies, etc.)
- [x] Ensure the Full Name of the tree node was constructed as expected

(For a video demonstrating the Issue see https://github.com/specify/specify7/issues/5074#issue-2393357073)

**whiteSpaceSensitive form attribute**
- Identify any non-relationship field on any table which is of type `text` or `String`
- On the Form Definition for the table, add the field to the definition (if needed) and add `whiteSpaceSensitive=true` to the initialize attributes of the field (example label and cell provided below)
- [x] Navigate to a DataEntry form of the table and enter a value which would normally be trimmed by Specify (such as `    test   `) and ensure the value remains as-is after the form is saved

No longer gets trimmed, everything looks good 

https://github.com/specify/specify7/assets/150940840/ded1f6ee-5048-40ae-8f8c-ecd99cfe7192

**Review by alesan99 — APPROVED** [review]
**fullNameSeparator**
- Make any edit to a TreeDefItem which is included in the fullName and save
  - For example, if Genus is edited then prior to this PR any `Genus Species` Taxon records would become `GenusSpecies` 
- In the TreeViewer or QueryBuilder, identify a node at a low level (Species, SubSpecies, etc.)
- [x] Ensure the Full Name of the tree node was constructed as expected

(For a video demonstrating the Issue see https://github.com/specify/specify7/issues/5074#issue-2393357073)

**whiteSpaceSensitive form attribute**
- Identify any non-relationship field on any table which is of type `text` or `String`
- On the Form Definition for the table, add the field to the definition (if needed) and add `whiteSpaceSensitive=true` to the initialize attributes of the field (example label and cell provided below)
- [x] Navigate to a DataEntry form of the table and enter a value which would normally be trimmed by Specify (such as `    test   `) and ensure the value remains as-is after the form is saved

Checks out 👌👌

image

image

File: specifyweb/frontend/js_src/lib/components/DataModel/schemaOverrides.ts [hunk 1]
```diff
@@ -29,14 +29,18 @@ export type TableConfigOverwrite =
...
-export type FieldConfigOverwrite =
-  // Makes a required field optional
-  | 'optional'
-  | 'required'
-  // Removes a field from the mapper (but not from Query Builder)
-  | 'readOnly'
-  // Hides a field. If it was required, it is made optional
-  | 'hidden';
+type FieldConfigOverwrite = Partial<{
+  readonly accessibility:
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/DataModel/schemaOverrides.ts:33]: (optional) accessibility name is fine, but `visibility` might be good too

File: specifyweb/frontend/js_src/lib/components/DataModel/specifyField.ts [hunk 2]
```diff
@@ -250,9 +254,20 @@ export class LiteralField extends FieldBase {
 
   public readonly isRelationship: false = false;
 
+  // Indicates white space should not be ignored in the field
+  // eslint-disable-next-line functional/prefer-readonly-type
+  public whiteSpaceSensitive: boolean;
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/DataModel/specifyField.ts:259]: add `readonly`?

**Review by maxpatiiuk — APPROVED** [review]
great job ❤️
we needed this

File: specifyweb/frontend/js_src/lib/components/DataModel/specifyField.ts (hunk 2 above)
Comment (melton-jason) [reply specifyweb/frontend/js_src/lib/components/DataModel/specifyField.ts]: Fixed in https://github.com/specify/specify7/pull/5076/commits/baeb3725f17fed8f57bcd817ab31a2724c05fcc4

**Review by emenslin — APPROVED** [review]
### Testing instructions

**fullNameSeparator**
- Make any edit to a TreeDefItem which is included in the fullName and save
  - For example, if Genus is edited then prior to this PR any `Genus Species` Taxon records would become `GenusSpecies` 
- In the TreeViewer or QueryBuilder, identify a node at a low level (Species, SubSpecies, etc.)
- [x] Ensure the Full Name of the tree node was constructed as expected

(For a video demonstrating the Issue see https://github.com/specify/specify7/issues/5074#issue-2393357073)

**whiteSpaceSensitive form attribute**
- Identify any non-relationship field on any table which is of type `text` or `String`
- On the Form Definition for the table, add the field to the definition (if needed) and add `whiteSpaceSensitive=true` to the initialize attributes of the field (example label and cell provided below)
- [x] Navigate to a DataEntry form of the table and enter a value which would normally be trimmed by Specify (such as `    test   `) and ensure the value remains as-is after the form is saved

👍

**Review by Areyes42 — APPROVED** [review]
### Testing instructions

**fullNameSeparator**

-   Make any edit to a TreeDefItem which is included in the fullName and save
    -   For example, if Genus is edited then prior to this PR any `Genus Species` Taxon records would become `GenusSpecies`
-   In the TreeViewer or QueryBuilder, identify a node at a low level (Species, SubSpecies, etc.)
-   [x] Ensure the Full Name of the tree node was constructed as expected

(For a video demonstrating the Issue see [#5074 (comment)](https://github.com/specify/specify7/issues/5074#issue-2393357073))

**whiteSpaceSensitive form attribute**

-   Identify any non-relationship field on any table which is of type `text` or `String`
-   On the Form Definition for the table, add the field to the definition (if needed) and add `whiteSpaceSensitive=true` to the initialize attributes of the field (example label and cell provided below)
-   [x] Navigate to a DataEntry form of the table and enter a value which would normally be trimmed by Specify (such as `test` ) and ensure the value remains as-is after the form is saved
***
Everything worked as expected with my tests, great job 👏

**Pull request merged**